docs/guide/data_migration.md

```diff
@@ -160,3 +307,313 @@ bin/console ibexa:migrations:kaliop:bulk-convert --recursive --input-directory=k
+In contrast with Kaliop migrations, actions provide you with ability to perform additional operations and extend
+the migration functionality. See [creating your own Actions](#creating-your-own-actions)
+
+## References
+
+References are key-value pairs necessary when one migration depends on another.
+
+Since some migrations generate object properties (like ID's) during their execution, which cannot be known in advance,
+references provide migrations with ability to use previously created object properties in further migrations.
+They can be subsequently used by passing them in their desired place with `reference:` prefix.
+
+Example below creates a Content object of type "folder", and stores it's location path as `"ref_path__folder__media"`.
+Then this reference is reused as part of a new role, as a limitation.
+
+```yaml
+-
+    type: content
+    mode: create
+    metadata:
+        contentType: folder
+        mainTranslation: eng-US
+        alwaysAvailable: true
+        section: 3
+        objectStates: {  }
+    location:
+        parentLocationId: 1
+        hidden: false
+        sortField: !php/const eZ\Publish\API\Repository\Values\Content\Location::SORT_FIELD_NAME
+        sortOrder: 1
+        priority: 0
+    fields:
+        -
+            fieldDefIdentifier: name
+            languageCode: eng-US
+            value: Media
+        # - ...
+    actions: {  }
+    references:
+        -
+            name: ref__content__folder__media
+            type: content_id
+        -
+            name: ref_location__folder__media
+            type: location_id
+        -
+            name: ref_path__folder__media
+            type: path
+
+-
+    type: role
+    mode: create
+    metadata:
+        identifier: foo
+    policies:
+        -
+            module: content
+            function: 'read'
+            limitations:
+                -
+                    identifier: Subtree
+                    values: ['reference:ref_path__folder__media']
+
+```
+
+By default, reference files are located in a separate directory `src/Migrations/Ibexa/references`
+(see [configuration reference](#configuration-reference)
+`ibexa_migrations.migration_directory` and `ibexa_migrations.references_files_subdir` options).
+
+Reference files are **NOT** loaded by default. A separate step (type: "reference", mode: "load", with filename as "value")
+is required. Similarly, saving a reference file is done using type: "reference", mode: "save" step, with filename.
+
+For example:
+```yaml
+-
+    type: reference
+    mode: load
+    filename: 'references.yaml'
+
+-
+    type: reference
+    mode: save
+    # You can also use 'references.yaml', in this case it will be overriden
+    filename: 'new_references.yaml'
+```
+
+!!! note
+
+    You don't need to save references if they are used in the same migration file.
+    References are stored in memory during migration, whether they are used or not.
+
+## How Ibexa Migrations work
```

Comment:
Not enthusiastic about this section. Assuming we have recipes for extending migrations (like with custom actions now, and presumably customer steps in future), does this section give the reader information they need to perform their tasks?

Reply:
If we remove this here, assuming this is not relevant for customers and only applies to our devs, then this part should become part of `ibexa/migrations` readme.

(not that anybody reads those)

My assumption for providing this is that our devs often use docs as reference or explanation. That's why I added it here. I'm ok with removal.